src/dispatch/experimental/durable/frame.c

```diff
@@ -120,44 +280,245 @@ static const char *get_type_name(PyObject *obj) {
+#elif PY_MINOR_VERSION == 12
+    Frame *frame = (Frame *)(struct _PyInterpreterFrame *)(gen_like->gi_iframe);
+#elif PY_MINOR_VERSION == 13
+    Frame *frame = (Frame *)(struct _PyInterpreterFrame *)(gen_like->gi_iframe);
+#endif
     assert(frame);
-    return (InterpreterFrame *)frame;
+    return frame;
+}
+
+static PyCodeObject *get_frame_code(Frame *frame) {
+#if PY_MINOR_VERSION == 10
+    PyCodeObject *code = frame->f_code;
+#elif PY_MINOR_VERSION == 11
+    PyCodeObject *code = frame->f_code;
+#elif PY_MINOR_VERSION == 12
+    PyCodeObject *code = frame->f_code;
+#elif PY_MINOR_VERSION == 13
+    PyCodeObject *code = (PyCodeObject *)frame->f_executable;
+#endif
+    assert(code);
+    return code;
+}
+
+static int get_frame_lasti(Frame *frame) {
+#if PY_MINOR_VERSION == 10
+    return frame->f_lasti;
+#elif PY_MINOR_VERSION == 11
+// https://github.com/python/cpython/blob/3.11/Include/internal/pycore_frame.h#L69
+    PyCodeObject *code = get_frame_code(frame);
+    assert(frame->prev_instr);
+    return (int)((intptr_t)frame->prev_instr - (intptr_t)_PyCode_CODE(code));
+#elif PY_MINOR_VERSION == 12
+// https://github.com/python/cpython/blob/3.12/Include/internal/pycore_frame.h#L77
+    PyCodeObject *code = get_frame_code(frame);
+    assert(frame->prev_instr);
+    return (int)((intptr_t)frame->prev_instr - (intptr_t)_PyCode_CODE(code));
+#elif PY_MINOR_VERSION == 13
+// https://github.com/python/cpython/blob/v3.13.0a5/Include/internal/pycore_frame.h#L73
+    PyCodeObject *code = get_frame_code(frame);
+    assert(frame->instr_ptr);
+    return (int)((intptr_t)frame->instr_ptr - (intptr_t)_PyCode_CODE(code));
+#endif
+}
+
+void set_frame_lasti(Frame *frame, int lasti) {
+#if PY_MINOR_VERSION == 10
+    frame->f_lasti = lasti;
+#elif PY_MINOR_VERSION == 11
+// https://github.com/python/cpython/blob/3.11/Include/internal/pycore_frame.h#L69
+    PyCodeObject *code = get_frame_code(frame);
+    frame->prev_instr = (_Py_CODEUNIT *)((intptr_t)_PyCode_CODE(code) + (intptr_t)lasti);
+#elif PY_MINOR_VERSION == 12
+// https://github.com/python/cpython/blob/3.12/Include/internal/pycore_frame.h#L77
+    PyCodeObject *code = get_frame_code(frame);
+    frame->prev_instr = (_Py_CODEUNIT *)((intptr_t)_PyCode_CODE(code) + (intptr_t)lasti);
+#elif PY_MINOR_VERSION == 13
+// https://github.com/python/cpython/blob/v3.13.0a5/Include/internal/pycore_frame.h#L73
+    PyCodeObject *code = get_frame_code(frame);
+    frame->instr_ptr = (_Py_CODEUNIT *)((intptr_t)_PyCode_CODE(code) + (intptr_t)lasti);
+#endif
+}
+
+static int get_frame_state(PyGenObject *gen_like) {
+#if PY_MINOR_VERSION == 10
+    Frame *frame = (Frame *)(gen_like->gi_frame);
+    if (!frame) {
+        return FRAME_CLEARED;
+    }
+    return frame->f_state;
+#elif PY_MINOR_VERSION == 11
+    return gen_like->gi_frame_state;
+#elif PY_MINOR_VERSION == 12
+    return gen_like->gi_frame_state;
+#elif PY_MINOR_VERSION == 13
+    return gen_like->gi_frame_state;
+#endif
+}
```

Comment:
Do you think it would be worth splitting these in different files instead of doing the macro checks each time?

```
frame.c
frame_310.h
frame_311.h
frame_312.h
frame_313.h
```

Then include them with:
```c
#if PY_MINOR_VERSION == 10
#include "frame_310.h"
#elif PY_MINOR_VERSION == 11
#include "frame_311.h"
#elif PY_MINOR_VERSION == 12
#include "frame_312.h"
#elif PY_MINOR_VERSION == 13
#include "frame_313.h"
#endif
```

Reply:
I like it, let me try it.

Reply:
See aaf9c26bfdd782c66be3d95a96f87e94f76f7264